core/services/bulletprooftxmanager/nonce_syncer.go

```diff
@@ -0,0 +1,406 @@
+func NewNonceSyncer(store *store.Store, config orm.ConfigReader, ethClient eth.Client) *NonceSyncer {
+	return &NonceSyncer{
+		store,
+		config,
+		ethClient,
+	}
+}
+
+// SyncAll syncs nonces for all keys in parallel
+//
+// This should only be called once, before the EthBroadcaster has started.
+// Calling it later is not safe and could lead to races.
+func (s NonceSyncer) SyncAll(ctx context.Context) (merr error) {
+	keys, err := s.store.SendKeys()
+	if err != nil {
+		return errors.Wrap(err, "NonceSyncer#fastForwardNoncesIfNecessary failed to get keys")
+	}
+
+	var wg sync.WaitGroup
+	var errMu sync.Mutex
+
+	wg.Add(len(keys))
+	for _, key := range keys {
+		go func(k models.Key) {
+			defer wg.Done()
+			if err := s.fastForwardNonceIfNecessary(ctx, k.Address.Address()); err != nil {
+				errMu.Lock()
+				defer errMu.Unlock()
+				merr = multierr.Combine(merr, err)
+			}
+		}(key)
+	}
+
+	wg.Wait()
+
+	return errors.Wrap(merr, "NonceSyncer#fastForwardNoncesIfNecessary failed")
+}
+
+func (s NonceSyncer) fastForwardNonceIfNecessary(ctx context.Context, address common.Address) error {
+	chainNonce, err := s.pendingNonceFromEthClient(ctx, address)
+	if err != nil {
+		return errors.Wrap(err, "GetNextNonce failed to loadInitialNonceFromEthClient")
+	}
+	if chainNonce == 0 {
+		return nil
+	}
+
+	localNonce, err := GetNextNonce(s.store.DB, address)
+	if err != nil {
+		return err
+	}
+	if chainNonce <= uint64(localNonce) {
+		return nil
+	}
+	logger.Warnw(fmt.Sprintf("NonceSyncer: address %s has been used before, either by an external wallet or a different Chainlink node. "+
+		"Local nonce is %v but the on-chain nonce for this account was %v. "+
+		"Please note that using the chainlink keys with an external wallet is NOT SUPPORTED and can lead to missed or stuck transactions. "+
+		"This Chainlink node will now take ownership of this address and may overwrite currently pending transactions",
+		address.Hex(), localNonce, chainNonce),
+		"address", address.Hex(), "localNonce", localNonce, "chainNonce", chainNonce)
+
+	// First fetch
+	//
+	// We have to get the latest block first and then fetch deeper blocks in a
+	// subsequent step, because otherwise we don't know what block numbers to
+	// query for.
+	// pending can be fetched in the initial request because it doesn't rely on
+	// any block number.
+	reqs := []rpc.BatchElem{
+		rpc.BatchElem{
+			Method: "eth_getBlockByNumber",
+			Args:   []interface{}{"pending", true},
+			Result: &models.Block{},
+		},
+		rpc.BatchElem{
+			Method: "eth_getBlockByNumber",
+			Args:   []interface{}{"latest", true},
+			Result: &models.Block{},
+		},
+	}
+
+	err = s.ethClient.BatchCallContext(ctx, reqs)
+	if err != nil {
+		return err
+	}
+
+	if reqs[1].Error != nil {
+		return errors.Wrap(reqs[1].Error, "latest block request returned error")
+	}
+	latestBlock, is := reqs[1].Result.(*models.Block)
+	if !is {
+		panic(fmt.Sprintf("invariant violation, expected %T but got %T", &models.Block{}, latestBlock))
```

Comment:
hm log broadcaster is printing errors and skipping items in case of wrong types - might be a bit safer than panic?

Reply:
In this case, it should be absolutely impossible that it has any other type. So panicking is ok here.

Reply:
👍🏼